### PR TITLE
[5.2] Laravel libsodium Hasing Library

### PR DESCRIPTION
**Pull request by scrothers**
First implementation of adding libsodium support to Laravel as an alternative hashing method.

https://paragonie.com/book/pecl-libsodium

As in PR #10198 I'm having a hard time getting Travis to build because of the sodium dependency. Any help is appreciated.

### CODE REVIEW

File: tests/Hashing/SodiumHasherTest.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+class BcryptHasherTest extends PHPUnit_Framework_TestCase
```
Comment (crynobone) [inline tests/Hashing/SodiumHasherTest.php]: should be `SodiumHasherTest` class.

File: src/Illuminate/Hashing/SodiumHasher.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @param  string  $value
+     * @param  array   $unusedOptions
+     * @return string
+     *
+     * @throws \RuntimeException
+     */
+    public function make($value, array $unusedOptions = [])
+    {
+        return \Sodium\crypto_pwhash_scryptsalsa208sha256_str(
```
Comment (GrahamCampbell) [inline src/Illuminate/Hashing/SodiumHasher.php]: you must always import things - laravel code never uses a leading `\` to get to a namespace

**Comment by taylorotwell** [comment]
Can you describe the advantages of using this over the native PHP bcrypt hashing?

**Comment by scrothers** [comment]
@taylorotwell It isn't to solve a "problem" with bcrypt, and I don't want to allude to that being the goal. Sure, there are some posts online which demonstrate a potential implementation issue with bcrypt, but I don't believe Laravel is susceptible to those types of issues.

The biggest reason I went and made this implementation is because libsodium is a derivative project from the NaCL project. The NaCL project created this library with the aim of creating a "standard" encryption, public key and hashing library which uses only the best practices of the various security methods available to end users. There are advantages in this library that create "better hashes" technically from the use of sha256/512 hashing mechanisms.

This is specifically in contrast to bcrypt, bcrypt uses blowfish as it's underlying hashing mechanism. Blowfish is however, not a hashing mechanism, but actually an encryption mechanism, meaning that the key to reverse the hash is derived from the original item being hashed. That's a /potential/ weak spot in the bcrypt hashing mechanism all together (albeit one that has never been exploited to my knowledge). 

So ultimately, the libsodium hashing technique provides a "real" hash, using the sha method which is true hashing. Sodium simply wraps these functions into the "best practices" so that they can be used effectively through a uniform API to the consuming application.

Advantages are using a standard library, if the user chooses, to do their hashing (and possibly encryption) in such a way that is compatible with other applications using these higher forms of encryption/hashing. 

It's an addition of convenience code essentially.

**Comment by taylorotwell** [comment]
We probably need to skip the test on Travis because the Sodium extension is not installed?

**Comment by scrothers** [comment]
I ideally would like to have the test included, but, I added an extension dependency to the test so that it skips if the extension is missing.

**Comment by taylorotwell** [comment]
On further thought, I'm going to go ahead and wait on this. It can be pretty easily released a package for the framework, and we don't really have a security problem with the current bcrypt hashing available in PHP 5.5+

**Comment by scrothers** [comment]
Sure thing, I'll leave the branch alive for anyone else who wants to play with it and I'll package it into a package later today :)

**Comment by scrothers** [comment]
I have gone ahead and turned this into a package located here in case anyone wants it:
https://github.com/scrothers/libsodium-laravel

**Pull request closed without merging**